plugins/tpm/fu-tpm-device.c

```diff
@@ -0,0 +1,232 @@
+/*
+ * Copyright (C) 2019 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <tss2/tss2_esys.h>
+
+#include "fu-tpm-device.h"
+#include "fu-tpm-eventlog-device.h"
+
+struct _FuTpmDevice {
+	FuUdevDevice		 parent_instance;
+};
+
+G_DEFINE_TYPE (FuTpmDevice, fu_tpm_device, FU_TYPE_UDEV_DEVICE)
+
+static void Esys_Finalize_autoptr_cleanup (ESYS_CONTEXT *esys_context)
+{
+	Esys_Finalize (&esys_context);
+}
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (ESYS_CONTEXT, Esys_Finalize_autoptr_cleanup)
+
+static gboolean
+fu_tpm_device_probe (FuUdevDevice *device, GError **error)
+{
+	FuTpmDevice *self = FU_TPM_DEVICE (device);
+
+	/* set the physical ID */
+	if (!fu_udev_device_set_physical_id (device, "tpm", error))
+		return FALSE;
+
+	/* already open, so don't touch */
+	g_object_set (self, "device-file", NULL, NULL);
+	return TRUE;
+}
+
+static gboolean
+fu_tpm_device_setup_event_log (FuTpmDevice *self, GError **error)
+{
+	gsize bufsz = 0;
+	g_autofree gchar *basename = NULL;
+	g_autofree gchar *fn = NULL;
+	g_autofree guint8 *buf = NULL;
+	g_autoptr(FuDevice) dev = NULL;
+
+	basename = g_path_get_basename (fu_device_get_physical_id (FU_DEVICE (self)));
+	fn = g_build_filename ("/sys/kernel/security", basename, "binary_bios_measurements", NULL);
+	if (!g_file_get_contents (fn, (gchar **) &buf, &bufsz, error))
+		return FALSE;
+	if (bufsz == 0) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INVALID_FILE,
+			     "failed to read data from %s", fn);
+		return FALSE;
+	}
+	dev = fu_tpm_eventlog_device_new (buf, bufsz, error);
+	if (dev == NULL)
+		return FALSE;
+	if (!fu_device_setup (dev, error))
+		return FALSE;
+	fu_device_add_child (FU_DEVICE (self), dev);
+	return TRUE;
+}
+
+static gboolean
+fu_plugin_tpm_get_uint32 (ESYS_CONTEXT *ctx, guint32 query, guint32 *val, GError **error)
+{
+	TSS2_RC rc;
+	g_autofree TPMS_CAPABILITY_DATA *capability = NULL;
+
+	rc = Esys_GetCapability (ctx, ESYS_TR_NONE, ESYS_TR_NONE, ESYS_TR_NONE,
+				 TPM2_CAP_TPM_PROPERTIES, query, 1, NULL, &capability);
+	if (rc != TSS2_RC_SUCCESS) {
+		g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
+			     "capability request failed for query %x", query);
+		return FALSE;
+	}
+	if (capability->data.tpmProperties.count == 0) {
+		g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
+			     "no properties returned for query %x", query);
+		return FALSE;
+	}
+	if (capability->data.tpmProperties.tpmProperty[0].property != query) {
+		g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
+			     "wrong query returned (got %x expected %x)",
+			     capability->data.tpmProperties.tpmProperty[0].property,
+			     query);
+		return FALSE;
+	}
+
+	if (val != NULL)
```

Comment:
I'd actually say this should be a `g_return_val_if_fail` for the function - is there actually a case that we want to be supporting sending a NULL pointer for `val`?

Reply:
Ya g_return_val_if_fail makes sense here.